pallets/asset-manager/src/lib.rs

```diff
@@ -394,6 +431,74 @@ pub mod pallet {
 		pub fn account_id() -> T::AccountId {
 			T::PalletId::get().into_account()
 		}
+
+		/// Strip AccountId from the multilocation.
+		pub(crate) fn extract_multilocation(
+			location: &MultiLocation,
+		) -> Option<<T::AssetConfig as AssetConfig<T>>::AssetLocation> {
+			let MultiLocation { parents, interior } = location;
+			// Currently, we only support X3 at most.
+			let striped_location = match interior {
+				// Send tokens back to relaychain.
+				Junctions::X1(Junction::AccountId32 { .. }) | Junctions::Here => {
+					MultiLocation::parent()
+				}
+				// Send native tokens to sibling chain.
+				Junctions::X2(Junction::Parachain(_para_id), Junction::AccountId32 { .. }) => {
+					// MultiLocation::new(*parents, Junctions::X1(Junction::Parachain(*para_id)))
+					interior.clone().split_last().0.into_exterior(*parents)
+				}
+				// Send tokens to specified pallet on sibling chain.
+				Junctions::X3(
+					Junction::Parachain(_para_id),
+					Junction::PalletInstance(_pallet_idx),
+					Junction::AccountId32 { .. },
+				) => interior.clone().split_last().0.into_exterior(*parents),
+				// Send tokens with specified token symbol to sibling chain.
+				Junctions::X3(
+					Junction::Parachain(_para_id),
+					Junction::GeneralKey(_key),
+					Junction::AccountId32 { .. },
+				) => interior.clone().split_last().0.into_exterior(*parents),
+				// For now, we don't support X4 or longer Junctions
+				_ => return None,
+			};
+
+			Some(<T::AssetConfig as AssetConfig<T>>::AssetLocation::from(
+				striped_location,
+			))
+		}
+	}
+
+	/// Check the multilocation is supported by calamari/manta.
+	impl<T: Config> Contains<MultiLocation> for Pallet<T> {
+		fn contains(location: &MultiLocation) -> bool {
+			if location.parent_count() != 1 {
+				return false;
+			}
```

Comment:
Line `475 - 478` is redundant, should be deleted.
The reason is we should take `Locations` in the `AssetManager` as ground truth. If we would like to filter what location we put in, we should filter in the entrance.

Reply:
If user input a wrong `parents`, we can avoid unnecessary reads.

Reply:
People need to pay for that anyway, so I don't think avoid storage reads is the real deal here. If we would like to have better sanity checking, this should be implemented in a separate trait in a more principled way. Say `MantaLocationChecking` or something like that.